include/swift/SIL/SILOpenedArchetypesTracker.h

```diff
@@ -0,0 +1,166 @@
+public:
+  typedef llvm::DenseMap<Type, SILValue> OpenedArchetypesDefsMap;
+  // Re-use pre-populated map if available.
+  SILOpenedArchetypesTracker(const SILFunction &F,
+                             SILOpenedArchetypesTracker &Tracker)
+      : F(F), OpenedArchetypesDefs(Tracker.OpenedArchetypesDefs) { }
+
+  // Re-use pre-populated map if available.
+  SILOpenedArchetypesTracker(const SILFunction &F,
+                             OpenedArchetypesDefsMap &OpenedArchetypesDefs)
+      : F(F), OpenedArchetypesDefs(OpenedArchetypesDefs) { }
+
+  // Use its own local map if no pre-populated map is provided.
+  SILOpenedArchetypesTracker(const SILFunction &F)
+      : F(F), OpenedArchetypesDefs(LocalOpenedArchetypesDefs) { }
+
+
+  const SILFunction &getFunction() const { return F; }
+
+  void addOpenedArchetypeDef(Type archetype, SILValue Def) {
+    assert(!getOpenedArchetypeDef(archetype) &&
+           "There can be only one definition of an opened archetype");
+    OpenedArchetypesDefs[archetype] = Def;
+  }
+
+  void removeOpenedArchetypeDef(Type archetype, SILValue Def) {
+    auto FoundDef = getOpenedArchetypeDef(archetype);
+    assert(FoundDef &&
+           "Opened archetype definition is not registered in SILFunction");
+    if (FoundDef == Def)
+      OpenedArchetypesDefs.erase(archetype);
+  }
+
+  // Return the SILValue defining a given archetype.
+  // If the defining value is not known, return an empty SILValue.
+  SILValue getOpenedArchetypeDef(Type archetype) const {
+    return OpenedArchetypesDefs.lookup(archetype);
+  }
+
+  const OpenedArchetypesDefsMap &getOpenedArchetypesDefs() const {
+    return OpenedArchetypesDefs;
+  }
+
+  // Register archetypes opened by a given instruction.
+  // Can be used to incrementally populate the mapping, e.g.
+  // if it is done when performing a scan of all instructions
+  // inside a function.
+  void registerOpenedArchetypes(const SILInstruction *I);
+
+  // Register opened archetypes whose definitions are referenced by
+  // the typedef operands of this instruction.
+  void registerUsedOpenedArchetypes(const SILInstruction *I);
+
+  // Unregister archetypes opened by a given instruction.
+  // Should be only called when this instruction is to be removed.
+  void unregisterOpenedArchetypes(const SILInstruction *I);
+
+  // Handling of instruction removal notifications.
+  bool needsNotifications() { return true; }
+
+  // Handle notifications about removals of instructions.
+  void handleDeleteNotification(swift::ValueBase *Value);
+
+  virtual ~SILOpenedArchetypesTracker() {
+    // Unregister the handler.
+    F.getModule().removeDeleteNotificationHandler(this);
+  }
+
+private:
+  /// The function whose opened archetypes are being tracked.
+  /// Used only for verification purposes.
+  const SILFunction &F;
+  /// Mapping from opened archetypes to their definitions.
+  OpenedArchetypesDefsMap &OpenedArchetypesDefs;
```

Comment:
This can be spelled "OpenedArchetypeDefs".
(Or just drop the "Def" everywhere and call it "OpenedArchetypes")